src/main/java/ch/njol/skript/expressions/ExprFormatTime.java

```diff
@@ -51,35 +58,44 @@
 		Skript.registerExpression(ExprFormatTime.class, String.class, ExpressionType.PROPERTY, "%dates% formatted [human-readable] [(with|as) %-string%]");
 	}
 
-	@SuppressWarnings("null")
+	@SuppressWarnings("NotNullFieldNotInitialized")
 	private SimpleDateFormat format;
 
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<? extends String> customFormat;
+
 	@Override
-	@SuppressWarnings("null")
+	@SuppressWarnings({"null", "unchecked"})
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		setExpr((Expression<? extends Date>) exprs[0]);
-		if (exprs[1] != null) {
-			if (!(exprs[1] instanceof Literal)) {
-				VariableString str = (VariableString) exprs[1];
-				if (!str.isSimple()) {
-					Skript.error("Date format must not contain variables!");
-					return false;
-				}
+		customFormat = (Expression<? extends String>) exprs[1];
+
+		if (customFormat instanceof Literal || (customFormat instanceof VariableString && ((VariableString) customFormat).isSimple())) {
+			try {
+				format = new SimpleDateFormat(customFormat.getSingle(null));
+			} catch (Exception e) {
+				Skript.error("Invalid date format: " + exprs[1]);
+				return false;
 			}
-			format = new SimpleDateFormat((String) exprs[1].getSingle(null));
-		} else {
+		} else if (customFormat == null) {
 			format = new SimpleDateFormat(defaultFormat);
 		}
-		
+
 		return true;
 	}
 
-
 	@Override
 	protected String[] get(Event e, Date[] source) {
 		return get(source, new Getter<String, Date>() {
 			@Override
 			public String get(Date date) {
+				if (customFormat != null) {
+					try {
+						format = new SimpleDateFormat(customFormat.getSingle(e));
+					} catch (Exception ex) {
+						return null;
+					}
+				}
 				return format.format(new java.util.Date(date.getTimestamp()));
 			}
 		});
```

Comment:
```suggestion
		SimpleDateFormat format = this.format;
		String formatString = customFormat.getSingle(e);
		if (format == null) {
			try {
				format = new SimpleDateFormat(formatString);
			} catch (IllegalArgumentException ex) {
				return null;
			}
		}
		
		return get(source, new Getter<String, Date>() {
			@Override
			public String get(Date date) {
				return format.format(new java.util.Date(date.getTimestamp()));
			}
		});
```

This makes it so it won't call getSingle multiple times, it won't catch other exceptions, it won't catch a possible IllegalArgumentException from getSingle, and it doesn't write back to the field for no reason